Do not move back the non-draining state in Quiesce

kkaneda opened the pull request:
@bdarnell , @spencerkimball , I believe this is what was proposed in #1040 . Could you take a look?

This allows tests that use multiTestContext to shut down gracefully even when multiple stores depend on one another.

Test flakiness with short `scanInterval` seems to be fixed. Some tests still didn't finish when `scanInterval` is very short (e.g., 10ms), but it seems that was due to starvation (e.g., scanner always starts a new scan instead of looking at a stopper's channel). Tests also sometimes get a panic, but that will be fixed by #1051 .

kkaneda committed:
Do not move back the non-draining state in Quiesce

This allows us to use Quiesce to coordinate the drain among stoppers
in multiTestContext. Please see #1040.

spencerkimball commented on storage/client_test.go:
```diff
@@ -156,11 +156,23 @@ func (m *multiTestContext) Start(t *testing.T, numStores int) {
+	var stoppers []*util.Stopper
+	stoppers = append(stoppers, m.clientStopper)
 	for _, s := range m.stoppers {
+		stoppers = append(stoppers, s)
```
You could replace lines 159-163 with:

    stoppers := append([]*util.Stopper{m.clientStopper}, m.stoppers...)

kkaneda replied:
Ah, that looks nicer. Done.

spencerkimball commented on util/stopper.go:
```diff
@@ -162,13 +162,20 @@ func (s *Stopper) SetStopped() {
 	defer s.mu.Unlock()
-	s.draining = 1
+	atomic.StoreInt32(&s.draining, 1)
+}
```
I don't think you should remove the call to `s.drain.Wait()` here. The name `Quiesce()` suggests that the tasks should in fact drain by the time it returns.

kkaneda replied:
Sure, fixed.

spencerkimball commented on util/stopper.go:
```diff
@@ -162,13 +162,20 @@ func (s *Stopper) SetStopped() {
 
-// Quiesce moves the stopper to state draining, waits until all tasks
-// complete, then moves back to non-draining state. This is used from
+// Quiesce moves the stopper to state draining. Unquiesce() needs to
```
Why does `Unquiesce()` need to be called before `Stop()`?

kkaneda replied:
You're right. This was unnecessary. I had somehow thought that it makes sense to move the state back to normal state before `Stop()`. Updated the comment.

spencerkimball commented on util/stopper.go:
```diff
@@ -162,13 +162,20 @@ func (s *Stopper) SetStopped() {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	s.draining = 1
+	atomic.StoreInt32(&s.draining, 1)
```
Now that you're correctly setting draining using an atomic op, there's nothing here that requires you get the mutex lock. Same thing goes for `Unquiesce()`.

kkaneda replied:
Hmm, updated, but I think `Quiesce()` still doesn't work in the following case:

0) `s.draining` is initially `0`.
1) `StartTask()` is called. It checks `s.draining`, but has not yet called `s.drain.Add(1)`.
2) `Quiesce()` is called. It sets `s.draining` and calls `s.drain.Wait()`. It finishes immediately.
3) `StartTask()` calls `s.drain.Add(1)`.
4) `Quiesce()` is called before all tasks have been drained.

bdarnell replied:
Put this in a comment in StartTask. It's not obvious that the lock is necessary since WaitGroup.Add is thread-safe. 

Alternately, call `s.drain.Add(1)` *before* checking `s.draining`. If draining is true, call `s.drain.Done()` and return false.

tamird replied:
If locking the mutex is necessary, why use an atomic op here?

kkaneda commented:
Thanks for the review, @spencerkimball . Updated the PR but I think there is a race that `Quiesce` finishes successfully while `StartTask()` is called. Will work on a fix and update the PR.

kkaneda commented:
Actually I've changed `StartTask()` to check `s.draining()` while having the mutex lock, and changed `Quiesce()` to still acquire the mutex lock. Does this make sense?

vivekmenezes commented:
Kenji, can you update the declaration of the lock mu to reflect what it is
protecting. thanks

On Tue, May 19, 2015 at 10:56 AM kkaneda <notifications@github.com> wrote:

> Actually I've changed StartTask() to check s.draining() while having the
> mutex lock, and changed Quiesce() to still acquire the mutex lock. Does
> this make sense?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1055#issuecomment-103536522>
> .
>

kkaneda commented:
Sure, @vivekmenezes ! Done.

bdarnell commented:
LGTM

tamird commented on storage/client_test.go:
```diff
@@ -156,11 +156,15 @@ func (m *multiTestContext) Start(t *testing.T, numStores int) {
-	m.clientStopper.Stop()
-	for _, s := range m.stoppers {
+	stoppers := append([]*util.Stopper{m.clientStopper}, m.stoppers...)
+	stoppers = append(stoppers, m.transportStopper)
```
does `transportStopper` have to be at the end, or it be inlined above? if it has to be at the end, can you add a comment explaining why?

bdarnell replied:
It had to be at the end before because the stores couldn't shut down cleanly if the transport was closed out from under them. Now that we have a two-phase quiesce and stop sequence, it may not be necessary (but I'm not sure).

kkaneda replied:
Thanks for catching. Yes, I think we can inline them. Fixed.

vivekmenezes replied:
LGTM

On Tue, May 19, 2015 at 10:39 PM kkaneda <notifications@github.com> wrote:

> In storage/client_test.go
> <https://github.com/cockroachdb/cockroach/pull/1055#discussion_r30667629>:
>
> > @@ -156,11 +156,15 @@ func (m *multiTestContext) Start(t *testing.T, numStores int) {
> >  }
> >
> >  func (m *multiTestContext) Stop() {
> > -	m.clientStopper.Stop()
> > -	for _, s := range m.stoppers {
> > +	stoppers := append([]*util.Stopper{m.clientStopper}, m.stoppers...)
> > +	stoppers = append(stoppers, m.transportStopper)
>
> Thanks for catching. Yes, I think we can inline them. Fixed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1055/files#r30667629>.
>

tamird commented on util/stopper.go:
```diff
@@ -50,8 +50,8 @@ type Stopper struct {
 	stopped  chan struct{}  // Closed when stopped completely
 	stop     sync.WaitGroup // Incremented for outstanding workers
-	mu       sync.Mutex     // Protects the fields below
 	draining int32          // 1 when Stop() has been called, updated atomically
```
please pardon my ignorance: why not use a bool here that is always updated under lock?

vivekmenezes commented on util/stopper.go:
```diff
@@ -106,6 +106,9 @@ func (s *Stopper) StartTask() bool {
 		defer s.mu.Unlock()
+		if atomic.LoadInt32(&s.draining) == 1 {
+			return false
+		}
```
I don't know why you're checking for draining again. It could always be draining after you check. Also the comments about the deadlocking above are scary because you could start draining just before you attempt to hold on to the lock.

kkaneda commented:
@tamird , I think the reason why we use an atomic op is to avoid the following deadlock:

1) `Stop()` is called and blocked at `s.stop.Wait()` after acquiring the lock.
2) `StartTask()` is called and attempts to acquire the lock. It is then blocked.
3) `Stop()` never finishes since the task doesn't call `SetStopped`.

@bdarnell , I went ahead and updated the code as you suggested in the comment. I think the code is better and `Stop()` now calls `Quiesce()` internally.

@vivekmenezes , you're right on the deadlock. I think acquiring the lock after checkign `s.draining` can still cause a deadlock. Updated the code as Ben suggested to address that.

Could you guys take another look? 

Thanks,
-Kenji

tamird commented on util/stopper.go:
```diff
@@ -97,18 +97,11 @@ func (s *Stopper) AddCloser(c Closer) {
 // Returns true if the task can be launched or false to indicate the
 // system is currently draining and the task should be refused.
 func (s *Stopper) StartTask() bool {
-	// Avoid locking when we're draining (which would deadlock
```
so this comment should be restored?

bdarnell replied:
No; there is no longer a risk of deadlock because Stop no longer acquires the lock until after Quiesce has finished.

bdarnell commented:
LGTM

kkaneda commented:
Sorry, hit the data race and needed to changed the code again.. It seems calling `WaitGroup.Add` while another gorountine is calling `WaitGroup.Done` caused data race. 

https://groups.google.com/forum/#!topic/golang-nuts/rAGsfCZkHP0

I've changed the code to simply use a mutex and a conditional variable...

tamird commented on util/stopper.go:
```diff
@@ -162,13 +160,20 @@ func (s *Stopper) SetStopped() {
+}
+
+// Unquiesce moves back to non-draining state. This is used from unittests.
+func (s *Stopper) Unquiesce() {
```
looks like this method is only in `TestStopperUnquiesce`, which only tests this method. Why not remove it?

kkaneda replied:
Sure, done. I think we can add it back easily when it is actually needed.

tamird commented on util/stopper_test.go:
```diff
@@ -163,6 +163,75 @@ func TestStopperRunWorker(t *testing.T) {
+			if s.StartTask() {
+				t.Error("expected StartTask to fail")
+			}
+			// Make the stoppers call Unquiesce().
```
I think this comment is erroneous

kkaneda replied:
Ah, yes.. Updated.

tamird commented on util/stopper_test.go:
```diff
@@ -163,6 +163,75 @@ func TestStopperRunWorker(t *testing.T) {
+
+		// Wait until StartTask() is called.
+		for _, sc := range startTaskDone {
+			<-sc
```
Could this be written to exercise each stopper in its own goroutine? This seems to serialize over the stoppers, which is probably not intended

kkaneda replied:
Hmm, this is actually the intended use case (`multiTestContext.Stop`)..

tamird replied:
Ah, LGTM

vivekmenezes replied:
LGTM

On Wed, May 20, 2015 at 9:13 AM Tamir Duberstein <notifications@github.com>
wrote:

> In util/stopper_test.go
> <https://github.com/cockroachdb/cockroach/pull/1055#discussion_r30699245>:
>
> > +			s.SetStopped()
> > +		}()
> > +	}
> > +
> > +	done := make(chan struct{})
> > +	go func() {
> > +		for _, s := range stoppers {
> > +			s.Quiesce()
> > +		}
> > +		for _, qc := range quiesceDone {
> > +			close(qc)
> > +		}
> > +
> > +		// Wait until StartTask() is called.
> > +		for _, sc := range startTaskDone {
> > +			<-sc
>
> Ah, LGTM
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1055/files#r30699245>.
>

vivekmenezes commented on util/stopper.go:
```diff
@@ -162,13 +160,13 @@ func (s *Stopper) SetStopped() {
-	s.draining = 0
+	s.draining = true
+	for s.numTasks > 0 {
+		s.drain.Wait()
```
appreciate a comment saying // unlocks and locks

kkaneda replied:
Sure, added the following comment.

```
// Unlock s.mu, wait for the signal, and lock s.mu.
```

tamird commented on util/stopper.go:
```diff
@@ -51,17 +50,20 @@ type Stopper struct {
 	stopped  chan struct{}  // Closed when stopped completely
 	stop     sync.WaitGroup // Incremented for outstanding workers
 	mu       sync.Mutex     // Protects the fields below
```
final nit: I believe you can remove this and embed `drain` by value, and then lock via `drain.L.Lock()`

kkaneda replied:
Hmm, I got a nil pointer dereference error. It seems we still need to call `sync.NewCond`. Also, it might be a bit confusing if `drain.L` is used to protect `closers`, etc... I'll keep the code as-is. (We could embed `sync.Mutex`, but I wasn't sure if it's better.)

bdarnell commented:
LGTM

kkaneda commented:
Thanks for all the reviews! Will merge after all tests passed.
